### PR TITLE
access log: add downstream transport failure reason

### PR DESCRIPTION
**Pull request by botengyao**
Signed-off-by: Boteng Yao <boteng@google.com>

access log: add downstream transport failure reason

Commit Message:
Additional Description:
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/listener_managers/listener_manager/active_stream_listener_base.cc
Comment (tyxia) [inline source/extensions/listener_managers/listener_manager/active_stream_listener_base.cc:115]: Here is possible that the connection is closed successfully without `transportFailure`, right? Do we want to treat that differently? i.e., not reporting error.

We can check` if (!connection_->transportFailureReason.empty())` first

File: test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc
Comment (tyxia) [inline test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc:780]: Yea, this is why I think we could check `!empty` first. IMO, reporting failureReason (even empty) is confusing for successful connection case.  WDYT?

What will be log_result here if you don't set `setDownstreamTransportFailureReason` above? I think it is ok we just don't have access log for successful connection case as this log is called failureReason anyways.

**Review by tyxia — COMMENTED** [review]
Thanks for working on this!

File: envoy/network/socket.h
Comment (botengyao) [reply envoy/network/socket.h:115]: Thanks! The reason is mainly that all the info in ConnectionInfo is related to downstream connection, we can even get the ssl_connection() it that class. downstream_transport_failure can be at that scope I think. I am okay to move, but let's see others' option.

**Review by botengyao — COMMENTED** [review]
Thanks for the review!

File: envoy/network/socket.h
Comment (botengyao) [reply envoy/network/socket.h:115]: Oh, I think I sticked to this `stream_info.downstreamAddressProvider().downstreamTransportFailureReason()` method. if ConnectionInfoProvider/Setter can be shared both upstream and downstream, it makes sense to move out from it. Thanks!

Comment (tyxia) [reply envoy/network/socket.h:115]: I know there are some downstream related code there. However, by nature, it is part of connection and it can be used by both upstream and downstream. right? See enableSettingInterfaceName() in that class which is used by upstream. 

Off top of my head, as I mentioned above, you can have a new class DownstreamInfo and create it in the `ServerConnectionImpl::ServerConnectionImpl` to strictly differentiate them, but I think it is overkill just for adding failure reason.

File: source/extensions/listener_managers/listener_manager/active_stream_listener_base.cc
Comment (tyxia) [reply source/extensions/listener_managers/listener_manager/active_stream_listener_base.cc:115]: Yea , I missed that you already check !empty in substitution_formatter.cc. Check here is not necessary

File: envoy/network/socket.h
Comment (tyxia) [reply envoy/network/socket.h:120]: No, there is no addictional copy if you `std::string, initializing std::string` as I mentioned above(last paragraph)
Also, see [go/totw/101](https://goto.google.com/totw/101)

Since you already stored `downstream_transport_failure_reason_` as string, you can just return std::string. ( neither const string& nor absl::string_view)

So, returning `absl::string_view` doesn't save anything.

File: envoy/network/socket.h
Comment (botengyao) [reply envoy/network/socket.h:115]: Agreed, Thanks!

File: envoy/network/socket.h
Comment (botengyao) [reply envoy/network/socket.h:120]: Oh I see, thanks so much! It may make sense to replace others.

Comment (tyxia) [reply envoy/network/socket.h:120]: Also, I do see lots of place returning `absl::string_view` in Envoy code base. I think it is fine to keep as it is.
I just want to mention this for your reference.

Comment (botengyao) [reply envoy/network/socket.h:120]: Good to know this C++17 stuff 😄, thanks!

File: api/envoy/data/accesslog/v2/accesslog.proto
Comment (htuch) [inline api/envoy/data/accesslog/v2/accesslog.proto:177]: Why does the downstream transport failure depend on upstream socket selection?

File: api/envoy/data/accesslog/v2/accesslog.proto
Comment (htuch) [inline api/envoy/data/accesslog/v2/accesslog.proto:73]: This is v2 API and frozen. Do you want to be updating v3?

File: docs/root/configuration/observability/access_log/usage.rst
Comment (htuch) [inline docs/root/configuration/observability/access_log/usage.rst:523]: Do we generate access logs for failed TLS handshakes?

Comment (botengyao) [reply docs/root/configuration/observability/access_log/usage.rst:523]: Yes, one possible error for TLS termination is: `DOWNSTREAM_TRANSPORT_FAILURE_REASON=TLS_error:_268435709:SSL_routines:"
                           "OPENSSL_internal:NO_COMMON_SIGNATURE_ALGORITHMS`

File: api/envoy/data/accesslog/v2/accesslog.proto
Comment (botengyao) [reply api/envoy/data/accesslog/v2/accesslog.proto]: Thanks! Yeah, V3 should be used.

File: api/envoy/data/accesslog/v2/accesslog.proto
Comment (botengyao) [reply api/envoy/data/accesslog/v2/accesslog.proto]: Ops, thanks for catching this. It depends on listener format setting I think.

**Review by tyxia — COMMENTED** [review]
Overall LGTM, just some comments. Thanks!

**Review by htuch — COMMENTED** [review]
/lgtm api

File: docs/root/configuration/observability/access_log/usage.rst
Comment (yanavlasov) [reply docs/root/configuration/observability/access_log/usage.rst:523]: Note, that the DOWNSTREAM_TRANSPORT_FAILURE_REASON is valid only in the listener access log. The HTTP or TCP access logs would always observe empty DOWNSTREAM_TRANSPORT_FAILURE_REASON.

I think it needs to be explained in the doc to reduce possible confusion.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: docs/root/configuration/observability/access_log/usage.rst
Comment (botengyao) [reply docs/root/configuration/observability/access_log/usage.rst:523]: Thanks! done!

File: source/common/formatter/substitution_formatter.cc
Comment (botengyao) [reply source/common/formatter/substitution_formatter.cc:1484]: Thanks!

File: test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc
Comment (botengyao) [reply test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc]: Done.

File: source/common/stream_info/stream_info_impl.h
Comment (tyxia) [inline source/common/stream_info/stream_info_impl.h:384]: Explicit `absl::string_view()`  conversion is redundant. The implicit conversion from string to string_view should be triggered

File: test/mocks/stream_info/mocks.cc
Comment (tyxia) [inline test/mocks/stream_info/mocks.cc:192]: Ditto. 
The key here is returning live value of `downstream_transport_failure_reason_`, string_view conversion here is redundant.

Alternatively you can use  `ReturnPointee` to return the live value, but it is up to you.
```
 ON_CALL(*this, downstreamTransportFailureReason())
      .WillByDefault(ReturnPointee(&downstream_transport_failure_reason_));
```

**Review by tyxia — APPROVED** [review]
LGTM, modulo two nits.

Thanks for your patience!

File: source/common/stream_info/stream_info_impl.h
Comment (botengyao) [reply source/common/stream_info/stream_info_impl.h:384]: My initial thinking is same,  but it will fail clang-asan check without this.

Comment (botengyao) [reply source/common/stream_info/stream_info_impl.h]: Thanks, changed back. The mock make me think somewhere wired.

File: test/mocks/stream_info/mocks.cc
Comment (botengyao) [reply test/mocks/stream_info/mocks.cc]: Thanks, done.

**Comment by botengyao** [comment]
Hi All, PTAL, Thanks!

**Review by tyxia — DISMISSED** [review]
LGTM, Thanks

File: api/envoy/data/accesslog/v3/accesslog.proto
Comment (yanavlasov) [inline api/envoy/data/accesslog/v3/accesslog.proto:221]:
```suggestion
  // downstream transport socket. Common TLS failures are in :ref:`TLS trouble shooting <arch_overview_ssl_trouble_shooting>`.
```

**Review by yanavlasov — CHANGES_REQUESTED** [review]
LGTM with a small documentation correction.

/wait

File: api/envoy/data/accesslog/v3/accesslog.proto
Comment (botengyao) [reply api/envoy/data/accesslog/v3/accesslog.proto]: Done, thanks!

**Review by yanavlasov — APPROVED** [review]

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by botengyao** [comment]
Thanks all for the review! Hi @envoyproxy/senior-maintainers, PTAL and can we get this merged? Thanks

**Review by htuch — APPROVED** [review]

**Comment by XinyiZhangAws** [comment]
Since Envoy does have attribute upstream.transport_failure_reason, do we have a plan to add downstream_transport_failure_reason as [connection attributes](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/attributes.html#connection-attributes) as well?

**Pull request merged**
